fem/restriction.hpp

```diff
@@ -104,10 +100,37 @@ class L2ElementRestriction : public Operator
    void FillJAndData(const Vector &ea_data, SparseMatrix &mat) const;
 };
 
+/// Base class for operators that extracts Face degrees of freedom.
+/** Objects of this type are typically created and owned by FiniteElementSpace
+    objects, see FiniteElementSpace::GetFaceRestriction(). */
+class FaceRestriction
+{
+protected:
+   int height;
+   int width;
+public:
+
+   FaceRestriction(): height(0), width(0) { }
+
+   FaceRestriction(int h, int w): height(h), width(w) { }
+
+   virtual ~FaceRestriction() { }
+
+   /** Extract the face degrees of freedom from @a x. */
+   virtual void Mult(const Vector &x, Vector &y) const = 0;
+
+   /** Add face degrees of freedom @a x to the element degrees of freedom @a y.*/
+   virtual void AddMultTranspose(const Vector &x, Vector &y) const = 0;
```

Comment:
Could these descriptions be expanded to include the information needed to use this interface (e.g. what is expected of "x": number of entries, how they're packed, ... )?

I think `x` is expected to be an "L-vector" and `y` is a (face) "E-vector" in `Mult()` and vice versa in `AddMultTranspose()`, but I'm not sure.

A more complete high-level description could be included on the class definition too.

Reply:
I pushed some documentation, I added a lot of documentation to this in #2195 .